pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -38,19 +42,132 @@ type Packings struct {
+		packings = append(packings, &Packings{podsPacked, instances})
+	}
+	return packings, nil
+}
+
+type packingEstimator struct {
+	isPodPacked map[*v1.Pod]bool
+}
+
+func (p *packingEstimator) calculatePackingForAllTypes(pods []*v1.Pod) ([]*v1.Pod, []string) {
+	previousPacking := []*v1.Pod{}
+	instanceTypesSelected := []*nodeCapacity{}
+	// Get all available instance types reverse sorted by their capacity, for every instance
+	// try to fit as many pods as possible. return the list of instances with
+	// highest pods packed and the pods.
+
+	// TODO add filters
+	// TODO reserve (Kubelet+ daemon sets) overhead for instance types
+	// TODO count number of pods created on an instance type
+	for _, instanceOption := range p.getInstanceTypes("") {
+		// check how many pods we can fit on this instance type
+		packings, err := p.calculatePackingForInstance(instanceOption, pods)
+		if err != nil {
+			zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instanceOption.instanceType, err)
+			continue
+		}
+		if len(packings) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if len(packings) == len(previousPacking) &&
+			podsMatch(previousPacking, packings) {
+			instanceTypesSelected = append(instanceTypesSelected, instanceOption)
+		} else if len(packings) > len(previousPacking) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			previousPacking = packings
+			instanceTypesSelected = []*nodeCapacity{instanceOption}
+		}
+	}
+	instanceTypeNames := []string{}
+	for _, instance := range instanceTypesSelected {
+		instanceTypeNames = append(instanceTypeNames, instance.instanceType)
+	}
+	return previousPacking, instanceTypeNames
+}
+
+func (p *packingEstimator) calculatePackingForInstance(instance *nodeCapacity, podList []*v1.Pod) ([]*v1.Pod, error) {
+	packing := []*v1.Pod{}
+	// start with the smallest pod based on resources requested
+	for _, pod := range podList {
+		if p.isPodPacked[pod] {
+			continue
+		}
+		cpu := calculateCPURequested(pod)
+		memory := calculateMemoryRequested(pod)
+		if err := instance.reserveCapacity(cpu, memory); err != nil {
+			if errors.Is(err, ErrInsufficientCapacity) {
+				// If we can't pack this pod, we can't do any better return the
+				// packings we have already calculated
+				break
+			}
+			return nil, fmt.Errorf("reserve capacity failed %w", err)
+		}
+		packing = append(packing, pod)
+	}
+	instance.utilizedCapacity = v1.ResourceList{
+		v1.ResourceCPU:    resource.MustParse("0"),
+		v1.ResourceMemory: resource.MustParse("0"),
+	}
+	return packing, nil
+}
+
+func podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podSeen := map[*v1.Pod]struct{}{}
+	for _, pod := range first {
+		podSeen[pod] = struct{}{}
```

Comment:
It is hard to say without a comment explaining what the semantics of `podsMatch` are, but this might not be working - you might need to hash the structure (https://github.com/mitchellh/hashstructure as used in another recent PR) rather than use the pointer?

Reply:
I have two list of pods, and need to compare both the list if it's the same set of pods. I have a comment from where `podsMatch` is called. 
It works in this case because the pointer value for the pod object is not changing and is used as a key for the map.

Reply:
But are you guaranteed that the two slices (`first` and `second`) don't have different copies of otherwise equal `v1.Pod` structs in them? (Also, even if that is true today, is it possible someone could change the behavior and not realize they're breaking this function?)

Reply:
I see your point, in my case it will be the same but if someone else might break this. I have changed it to use `namespace/podname` as the key which is guaranteed to be unique in the cluster.